File: java-app/src/main/java/com/example/javaapp/model/TaskStatus.java
package com.example.javaapp.model;

public enum TaskStatus {
    PENDING("Pending"),
    IN_PROGRESS("In Progress"),
    COMPLETED("Completed"),
    BLOCKED("Blocked"),
    CANCELLED("Cancelled");

    private final String displayName;

    TaskStatus(String displayName) {
        this.displayName = displayName;
    }

    public String getDisplayName() {
        return displayName;
    }

    public boolean isTerminal() {
        return this == COMPLETED || this == CANCELLED;
    }

    public boolean canTransitionTo(TaskStatus newStatus) {
        if (this == newStatus) return true;

        switch (this) {
            case PENDING:
                return newStatus == IN_PROGRESS || newStatus == BLOCKED || newStatus == CANCELLED;
            case IN_PROGRESS:
                return newStatus == COMPLETED || newStatus == BLOCKED || newStatus == CANCELLED;
            case BLOCKED:
                return newStatus == PENDING || newStatus == CANCELLED;
            case COMPLETED:
            case CANCELLED:
                return false;  // Terminal states
            default:
                return false;
        }
    }
}

File: js-app/models/TaskStatus.js
// Equivalent to Java Enum with transition rules
const TaskStatus = Object.freeze({
    PENDING: { name: 'PENDING', displayName: 'Pending' },
    IN_PROGRESS: { name: 'IN_PROGRESS', displayName: 'In Progress' },
    COMPLETED: { name: 'COMPLETED', displayName: 'Completed' },
    BLOCKED: { name: 'BLOCKED', displayName: 'Blocked' },
    CANCELLED: { name: 'CANCELLED', displayName: 'Cancelled' }
});

// Get status by name
const getStatusByName = (name) => {
    const upperName = name.toUpperCase().replace(' ', '_').replace('-', '_');
    return Object.values(TaskStatus).find(s => s.name === upperName) || TaskStatus.PENDING;
};

// Check if status is terminal (completed or cancelled)
const isTerminal = (status) => {
    return status === TaskStatus.COMPLETED || status === TaskStatus.CANCELLED;
};

// Check if transition is valid (same logic as Java)
const canTransitionTo = (currentStatus, newStatus) => {
    if (currentStatus === newStatus) return true;

    switch (currentStatus) {
        case TaskStatus.PENDING:
            return newStatus === TaskStatus.IN_PROGRESS ||
                   newStatus === TaskStatus.BLOCKED ||
                   newStatus === TaskStatus.CANCELLED;
        case TaskStatus.IN_PROGRESS:
            return newStatus === TaskStatus.COMPLETED ||
                   newStatus === TaskStatus.BLOCKED ||
                   newStatus === TaskStatus.CANCELLED;
        case TaskStatus.BLOCKED:
            return newStatus === TaskStatus.PENDING ||
                   newStatus === TaskStatus.CANCELLED;
        case TaskStatus.COMPLETED:
        case TaskStatus.CANCELLED:
            return false;
        default:
            return false;
    }
};

const getAllStatuses = () => Object.values(TaskStatus);

export { TaskStatus, getStatusByName, isTerminal, canTransitionTo, getAllStatuses };